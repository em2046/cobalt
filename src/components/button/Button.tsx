import ClassNameBuilder from '../../utils/class-name-builder';
import * as React from 'react';
import Icon from '../icon';
import { IconType } from '../icon/icon-sheet-generated';

export type ButtonType = 'default' | 'primary' | 'danger';
export type ButtonHtmlType = 'button' | 'submit' | 'reset';
export type ButtonSize = 'large' | 'small' | 'default';

export interface CustomButtonProps {
  icon?: IconType;
  className?: string;
  children?: React.ReactNode;
  type?: ButtonType;
  size?: ButtonSize;
  disabled?: boolean;
  loading?: boolean;
  htmlType?: ButtonHtmlType;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
}

type NativeButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement>;
export type ButtonProps = NativeButtonProps & CustomButtonProps;

export default function Button(props: ButtonProps) {
  const prefixClass = 'cobalt-button';
  const {
    icon,
    className,
    children,
    type = 'default',
    size,
    disabled,
    loading,
    htmlType,
    onClick,
    ...rest
  } = props;

  let iconType = loading ? 'loading' : icon;
  let iconNode = iconType ? <Icon type={iconType} /> : null;
  let isIconOnly = iconNode && !children;
  let classNameBuilder = new ClassNameBuilder(
    prefixClass,
    className,
    `${prefixClass}-${type}`,
    {
      [`${prefixClass}-${size}`]: size,
      [`${prefixClass}-loading`]: loading,
      [`${prefixClass}-icon-only`]: isIconOnly
    }
  );
  let classes = classNameBuilder.toString();

  function handleClick(e: React.MouseEvent<HTMLButtonElement, MouseEvent>) {
    if (disabled || loading) {
      return;
    }
    if (onClick) {
      onClick(e);
    }
  }

  return (
    <button
      className={classes}
      type={htmlType}
      disabled={disabled}
      onClick={handleClick}
      {...rest}
    >
      {iconNode}
      {children}
    </button>
  );
}
