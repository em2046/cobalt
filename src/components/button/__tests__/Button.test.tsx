import React, { useState } from 'react';
import renderer from 'react-test-renderer';
import Button from '../Button';

describe('Button', () => {
  it('renders without crashing', () => {
    const component = renderer.create(<Button>Button</Button>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders loading', () => {
    function LoadingButton() {
      const [loading, setLoading] = useState(false);

      return (
        <Button loading={loading} onClick={() => setLoading(!loading)}>
          Button
        </Button>
      );
    }

    const component = renderer.create(<LoadingButton />);
    let tree = component.toJSON();
    expect(tree).not.toBeNull();
    expect(tree).toMatchSnapshot();

    if (tree) {
      tree.props.onClick();
    }

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders disabled', () => {
    function DisabledButton() {
      const [disabled, setDisabled] = useState(false);

      return (
        <Button disabled={disabled} onClick={() => setDisabled(!disabled)}>
          Button
        </Button>
      );
    }

    // disabled == false
    const component = renderer.create(<DisabledButton />);
    let tree = component.toJSON();
    expect(tree).not.toBeNull();
    expect(tree).toMatchSnapshot();

    // disabled == true
    if (tree) {
      tree.props.onClick();
    }

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // disabled == true
    if (tree) {
      tree.props.onClick();
    }
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders {0}, 0, {false}', () => {
    const componentSetZero = renderer.create(<Button>{0}</Button>);
    let treeSetZero = componentSetZero.toJSON();
    expect(treeSetZero).toMatchSnapshot();

    const componentZero = renderer.create(<Button>0</Button>);
    let treeZero = componentZero.toJSON();
    expect(treeZero).toMatchSnapshot();

    const componentFalse = renderer.create(<Button>{false}</Button>);
    let treeFalse = componentFalse.toJSON();
    expect(treeFalse).toMatchSnapshot();
  });
});
