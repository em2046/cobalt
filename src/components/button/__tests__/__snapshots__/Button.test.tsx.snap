// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders {0}, 0, {false} 1`] = `
<button
  className="cobalt-button cobalt-button-default"
  onClick={[Function]}
>
  0
</button>
`;

exports[`Button renders {0}, 0, {false} 2`] = `
<button
  className="cobalt-button cobalt-button-default"
  onClick={[Function]}
>
  0
</button>
`;

exports[`Button renders {0}, 0, {false} 3`] = `
<button
  className="cobalt-button cobalt-button-default"
  onClick={[Function]}
/>
`;

exports[`Button renders disabled 1`] = `
<button
  className="cobalt-button cobalt-button-default"
  disabled={false}
  onClick={[Function]}
>
  Button
</button>
`;

exports[`Button renders disabled 2`] = `
<button
  className="cobalt-button cobalt-button-default"
  disabled={true}
  onClick={[Function]}
>
  Button
</button>
`;

exports[`Button renders disabled 3`] = `
<button
  className="cobalt-button cobalt-button-default"
  disabled={true}
  onClick={[Function]}
>
  Button
</button>
`;

exports[`Button renders loading 1`] = `
<button
  className="cobalt-button cobalt-button-default"
  onClick={[Function]}
>
  Button
</button>
`;

exports[`Button renders loading 2`] = `
<button
  className="cobalt-button cobalt-button-default cobalt-button-loading"
  onClick={[Function]}
>
  <span
    className="cobalt-icon cobalt-icon-spin"
  >
    <svg
      fill="currentColor"
      height="1em"
      viewBox="0 0 32 32"
      width="1em"
    >
      loading.svg
    </svg>
  </span>
  Button
</button>
`;

exports[`Button renders loading 3`] = `
<button
  className="cobalt-button cobalt-button-default cobalt-button-loading"
  onClick={[Function]}
>
  <span
    className="cobalt-icon cobalt-icon-spin"
  >
    <svg
      fill="currentColor"
      height="1em"
      viewBox="0 0 32 32"
      width="1em"
    >
      loading.svg
    </svg>
  </span>
  Button
</button>
`;

exports[`Button renders without crashing 1`] = `
<button
  className="cobalt-button cobalt-button-default"
  onClick={[Function]}
>
  Button
</button>
`;
